### DISCUSSION TITLE
Add "depth of discharge" and "inverter operation mode" entities for Goodwe inverters

### DISCUSSION THREAD

**Discussion by rinners** [slipx06/sunsynk-power-flow-card#235] [Goodwe]
This card is great, I get a little excited every time I see a new update. Big fist pump.

The only things I can currently think that I would want to add, for Goodwe inverters (not sure if others would benefit), is a way to display the depth of discharge and inverter operation mode entities, as I use these frequently and they currently cluttering up my dashboard elsewhere.

**Comment by slipx06** [comment]
😄 Thanks. If you look at the [example](https://slipx06.github.io/sunsynk-power-flow-card/examples/goodwe.html
), the template sensor for shutdown soc is based on the depth of discharge so it should be possible to calculate it from this which is shown on the card. There was also some discussion here https://github.com/slipx06/sunsynk-power-flow-card/issues/144 around `offgrid shutdown soc` and how you can display this.


No sure what you mean by operation mode entities? Depending on which inverter model you set (goodwe_gridmode or goodwe) a number of status messages are displayed

```
    goodwe_gridmode: {
        idle: {states: ['0', 'idle'], color: 'blue', message: localize('common.idle')},
        exporting: {states: ['1', 'exporting'], color: 'green', message: localize('common.exporting')},
        importing: {states: ['2', 'importing'], color: 'red', message: localize('common.importing')},
    },
    goodwe: {
        standby: {states: ['0', 'wait mode'], color: 'blue', message: localize('common.standby')},
        ongrid: {
            states: ['1', 'normal (on-grid)'], color: 'green', message: localize('common.ongrid')
        },
        offgrid: {
            states: ['2', 'normal (off-grid)'],
            color: 'green',
            message: localize('common.offgrid')
        },
        fault: {states: ['3', 'fault mode'], color: 'red', message: localize('common.fault')},
        flash: {states: ['4', 'flash mode'], color: 'yellow', message: localize('common.flash')},
        check: {states: ['5', 'check mode'], color: 'orange', message: localize('common.check')},
    }
```

**Reply by rinners** [reply]
I did see that thanks. I'm not familiar with other inverters, but for the Goodwe GW5048D-ES (and probably other Goodwe) the operation modes are:

![2024-02-13 08_15_17-Window](https://github.com/slipx06/sunsynk-power-flow-card/assets/48795308/75742d07-6b18-4da2-b2d7-64afd6d5c0f0)

General mode - will discharge battery down to the level set by the `depth of discharge` given enough load. I find this useful for discharging the batteries before the sun rises so I can make use of more PV charging the batteries. I then set the depth of discharge to 0% once there is insufficient PV to cover loads and reserve the battery to keep the lights on at night.

Backup mode - Prioritizes PV + about 400W grid to charge the batteries. No discharge happens until grid feed is unavailable and will then discharge to the `offgrid shutdown soc`. _This offgrid SoC setting is only visible in the Goodwe apps, not in the HA integration._ Depth of discharge setting is not used at all in this mode.

Off grid mode - I never use, but seems self explanatory. I assume it will never draw from the grid and will use `offgrid shutdown soc`

Eco mode - I assume is similar to those 6 time slots I've seen in the Synsynk config only I think you can set as many as needed. The only use I've had for this mode so far is to charge from the grid at more than the 400W limit that backup mode imposes. I rarely use this and it is only possible using the app on my inverter. Even if I had set up some charging window (say 00:00 - 23:59 @ 100% rate of charge) it gets wiped when I change mode in HA. 

So what I would find useful:

A: Replace that shutdown SoC with the depth of discharge entity. 

![2024-02-13 08_20_43-Window](https://github.com/slipx06/sunsynk-power-flow-card/assets/48795308/b9343052-2d68-4963-9c8b-0bad794f64e6)

B: Have the Operation Mode displayed. It could even replace the inverter status (as this is pretty much handled by changing the colours of the grid section) or in the empty space to the right of the inverter.

![2024-02-13 08_21_02-Window](https://github.com/slipx06/sunsynk-power-flow-card/assets/48795308/8f5150de-4454-416a-8a83-50f06b496b61)

Not sure what others think? I'll try go through that discussion a bit later
